core/domain/classifier/base_classification_algorithm.py

```diff
@@ -0,0 +1,153 @@
+# coding: utf-8
+#
+# Copyright 2015 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Base class of classification algorithms for free-form text answers"""
+
+from abc import abstractmethod
+import copy
+
+
+class BaseClassificationAlgorithm(object):
+    """A base class for classifiers that uses supervised learning to match
```

Comment:
Oh, sure! I think what we can do here is not to put string classifier's documentation here and think of the BaseClassificationAlgorithm class as a template for "any" classifier. For example, in the [load_examples](https://github.com/Andimeo/oppia/blob/14987c24c27cd754d2f2864556e42209259370aa/core/domain/classifier/base_classification_algorithm.py#L140) method, the argument 'examples' can be of the form [data, [labels]] where data can be anything (a string in the case of String classifier or some other data structure alternatively). So you can probably say "_examples: list of 'examples'. Each example is represented by a 2-element list. The first item of the list is the single training data, and the second item is a list of labels that the data should be matched"_. Then in the actual StringClassifier class, you can give a more concrete example like: 
_training_examples = [
    ['i eat fish and vegetables', ['food']],
    ['fish are pets', ['pets']],
    ['my kitten eats fish', ['food', 'pets']]
                ]_
I think the same goes for the other part of documentation as well. We can keep the string classifier specific documentation to the StringClassifer class. The documentation in this class should be a generic indicator of what the derived class should do. Please let me know if you'd like  me to elaborate further.